Add convenience wrapper for disk uploads to SDK

wfchandler opened the pull request:
From a user perspective uploading a disk is a single operation, but doing so via the SDK is a very involved process, involving stringing together multiple API calls with any interruption leaving the disk in a state that cannot be directly removed.

Add a new `extras` module to the SDK that will contain convenience wrappers around actions that functions that are overly complicated. Create a `disk_import` function in `extras` that will perform a disk upload as a single step, and attempts to cleanup partially imported disks on failure. We make the `disk upload` CLI command a thin wrapper around the new SDK function, responsible only for cancellation and progress. We manually recreate the builder interface, as creating a procmacro to do this would require creating a new crate. If/when `extras` continues to grow we should reconsider this.

`disk_import` also give callers access to a `DiskImportHandle` with a `watch` channel with the current number of bytes uploaded for use in progress tracking, and provides a `cancel` function that allows users to gracefully stop the disk upload. A disk that has been successfully created will not be removed, only those that are partially imported.

The `cancel` function uses `tokio::select` to cancel the `Future` of the upload task. This should be safe in our in our case no locks are held across `await` points. No operations need to be safe to restart as we only call `tokio::select` a single time, they will be dropped immediately and losing a message is not a problem.

We are forced to remove tests `test_disk_import_bulk_import_start_fail` and `test_disk_import_bulk_write_import_fail` as we now poll for disks in `Created` state, which may happen if a user immediately cancels a new request. This state is not eligible to be finalized, so we need to wait for the disk to progress to the next state, but `HttpMock` does not provide a way to programatically delete or update a mock after sending `n` responses. This means that the disk will always be returned as non-existing when querying, and cleanup will be skipped. I have opened https://github.com/alexliesenfeld/httpmock/pull/108 to add a `delete_after_n` method, but this seems unlikely to be merged. We may want to consider moving to `wiremock` in the long term, which provides this functionality.

wfchandler committed:
Add convenience wrapper for disk uploads to SDK

From a user perspective uploading a disk is a single operation, but
doing so via the SDK is a very involved process, involving stringing
together multiple API calls with any interruption leaving the disk in a
state that cannot be directly removed.

Add a new `extras` module to the SDK that will contain convenience
wrappers around actions that functions that are overly complicated.
Create a `disk_import` function in `extras` that will perform a disk
upload as a single step, and attempts to cleanup partially imported
disks on failure. We make the `disk upload` CLI command a thin wrapper
around the new SDK function, responsible only for cancellation and
progress. We manually recreate the builder interface, as creating a
procmacro to do this would require creating a new crate. If/when
`extras` continues to grow we should reconsider this.

`disk_import` also give callers access to a `DiskImportHandle` with a
`watch` channel with the current number of bytes uploaded for use in
progress tracking, and provides a `cancel` function that allows users
to gracefully stop the disk upload. A disk that has been successfully
created will not be removed, only those that are partially imported.

The `cancel` function uses `tokio::select` to cancel the `Future` of
the upload task. This should be safe in our in our case no locks are
held across `await` points. No operations need to be safe to restart as
we only call `tokio::select` a single time, they will be dropped
immediately and losing a message is not a problem.

We are forced to remove tests `test_disk_import_bulk_import_start_fail`
and `test_disk_import_bulk_write_import_fail` as we now poll for disks
in `Created` state, which may happen if a user immediately cancels a new
request. This state is not eligible to be finalized, so we need to wait
for the disk to progress to the next state, but `HttpMock` does not
provide a way to programatically delete or update a mock after sending
`n` responses. This means that the disk will always be returned as
non-existing when querying, and cleanup will be skipped. I have opened
https://github.com/alexliesenfeld/httpmock/pull/108 to add a
`delete_after_n` method, but this seems unlikely to be merged. We may
want to consider moving to `wiremock` in the long term, which provides
this functionality.

Co-authored-by: Adam Leventhal <ahl@oxide.computer>

wfchandler commented:
Demo of cancellation in action

https://github.com/user-attachments/assets/5f20d2c8-e1cc-453b-bb67-c0528ab020ee

david-crespo commented:
That is so great. "Cleaning up disk" is a little vague. I wonder if it's worth throwing a little more info in like the web console does, both before and after the main upload step. Like you could say "Creating disk xxxxxx", putting it in import mode, then the progress bar, and then when you cancel, the cleanup step makes more sense because it's undoing the steps you called out before. I'm a little on the fence because it's borderline too much detail, but I do like educating the user on what is actually taking place because it empowers them to debug and read the docs if something goes wrong.

wfchandler commented:
> That is so great. "Cleaning up disk" is a little vague. I wonder if it's worth throwing a little more info in like the web console does, both before and after the main upload step. Like you could say "Creating disk xxxxxx", putting it in import mode, then the progress bar, and then when you cancel, the cleanup step makes more sense because it's undoing the steps you called out before. I'm a little on the fence because it's borderline too much detail, but I do like educating the user on what is actually taking place because it empowers them to debug and read the docs if something goes wrong.

Listing the name makes sense, I've added that and replaced the debug formatting with quoted names throughout. Updated demo:

https://github.com/user-attachments/assets/bbc28967-c741-4c23-8bad-ef5a4a8d1126

augustuswm commented on sdk/Cargo.toml:
```diff
@@ -8,11 +8,13 @@ repository = "https://github.com/oxidecomputer/oxide.rs"
 clap = { workspace = true, optional = true }
 dirs = { workspace = true }
 futures = { workspace = true }
+indicatif = { workspace = true }
```
This looks to be unused now.

wfchandler replied:
Whoops, fixed in 7cc0419cc159d5b4a48a8d8c9bec1dddc7a53cb3

augustuswm commented on sdk/src/extras/disk.rs:
```diff
@@ -0,0 +1,711 @@
+            thread_ct: Option<usize>,
+        ) -> Result<
+            (
+                impl Future<Output = Result<()>> + 'a,
```
I still need to read through everything, but are we able to returned a typed error instead of an `anyhow` error? I think it is fine at the CLI level to return them, but `anyhow` errors are quite difficult to handle when getting them back from an sdk.

wfchandler replied:
Good call, I've moved to a custom error type.

augustuswm reviewed (commented):
Really appreciate this work! Will look through it, just a couple initial notes.

augustuswm commented on sdk/src/extras/disk.rs:
```diff
@@ -0,0 +1,711 @@
+            .client
+            .snapshot_view()
+            .project(&self.project)
+            .snapshot(NameOrId::Name(image_info.snapshot.clone()))
```
`Name` already implements `TryInto<NameOrId>` so you don't need to have the explicit variant here.

wfchandler replied:
Good catch, fixed in 7cc0419cc159d5b4a48a8d8c9bec1dddc7a53cb3

augustuswm commented on sdk/src/extras/disk.rs:
```diff
@@ -0,0 +1,711 @@
+            self.client
+                .disk_view()
+                .project(&self.project)
+                .disk(NameOrId::Name(self.disk.clone()))
```
Same as above.

wfchandler replied:
Fixed in 7cc0419cc159d5b4a48a8d8c9bec1dddc7a53cb3

augustuswm commented on sdk/src/extras/disk.rs:
```diff
@@ -0,0 +1,711 @@
+            .client
+            .disk_view()
+            .project(&self.project)
+            .disk(NameOrId::Name(self.disk.clone()))
```
Same as above.

wfchandler replied:
Fixed in 7cc0419cc159d5b4a48a8d8c9bec1dddc7a53cb3

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -93,461 +78,107 @@ pub struct CmdDiskImport {
-            return Err(e.into());
-        }
+            pb.finish_and_clear();
+            pb.println("Cleaning up disk. Press CTRL+C again to exit immediately.");
```
I would suggest we omit the "... exit immediately" text in that basically we don't want people to do that... and in my experience they don't need an invitation to ^C

wfchandler replied:
Ha, very true

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -0,0 +1,785 @@
+    ) -> Result<(), DiskImportError> {
+        self.check_for_existing_disk().await?;
+
+        let result = tokio::select! {
```
do we want `biased;` here? i.e. I think we want to prioritize cancelation over progress.

wfchandler replied:
Good catch

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -0,0 +1,785 @@
+        Ok(())
+    }
+
+    async fn import_disk(&self) -> Result<(), DiskImportError> {
```
```suggestion
    async fn do_disk_import(&self) -> Result<(), DiskImportError> {
```

What do you think? Hoping for a name that indicates this is the inner loop of `disk_import` rather than just reversing the two words

wfchandler replied:
Makes sense 👍

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -93,461 +78,107 @@ pub struct CmdDiskImport {
-                    break Err(e.into());
-                }
-            };
+        import_future.await?;
```
I'm not sure if this is going to be more complex or simpler, but what do you think of doing something like this:

```
let real_work_handle = tokio::spawn(import_future);

select! {
    // ^C thing
    // progress thing
}

real_work_handle.await?;
```

The difference I see is inlining the work that's spread between the progress bar and ctrl_c watch functions which don't seem to be particularly self-contained.

Perhaps it's about the same, but I wanted to float the idea.

wfchandler replied:
That'a bit cleaner, I combined the two tasks.

wfchandler replied:
Ended up reverting this as exiting the task when the progress bar completes causes us to stop listening for the cancel prior to the task completing, preventing users from canceling late in the request.

ahl reviewed (approved):
This looks great. I think the approach works.
